File: ocaml/lib/day4.ml
open Base
open Stdex

type range = (int * int)
type input = (range * range) list

let make_pair lst = (List.nth_exn lst 0, List.nth_exn lst 1)

let parse_range s = 
  String.split s ~on:'-' 
  |> List.map ~f:Int.of_string 
  |> make_pair

let parse_line s = 
  String.split s ~on:',' 
  |> List.map ~f:parse_range 
  |> make_pair

let parse_input lines = List.map ~f:parse_line lines

let contains ((min_a, max_a), (min_b, max_b)) =
  (min_a <= min_b && max_b <= max_a) ||
  (min_b <= min_a && max_a <= max_b)

let overlaps ((min_a, max_a), (min_b, max_b)) =
  (min_b <= min_a && min_a <= max_b) ||
  (min_b <= max_a && max_a <= max_b) ||
  (min_a <= min_b && max_b <= max_a) ||
  (min_b <= min_a && max_a <= max_b)

let part1 = List.count ~f:contains 
let part2 = List.count ~f:overlaps

File: haskell/day4.hs
module Day4 where

import Util
import Data.List
import qualified Aoc

parseRange :: String -> (Int, Int)
parseRange = listToPair . map read . splitList '-'

parse :: [String] -> [((Int, Int), (Int, Int))]
parse = map (listToPair . map parseRange . splitList ',')

isSorted xs = sort xs == xs

contains ((minA, maxA), (minB, maxB)) =
  isSorted [minA, minB, maxB, maxA] ||
  isSorted [minB, minA, maxA, maxB]

overlaps ((minA, maxA), (minB, maxB)) =
  isSorted [minB, minA, maxB] ||
  isSorted [minB, maxA, maxB] ||
  isSorted [minA, minB, maxB, maxA] ||
  isSorted [minB, minA, maxA, maxB]

part1 = length . filter contains
part2 = length . filter overlaps
run = Aoc.run 4 parse part1 part2

